core/src/main/java/de/bwaldvogel/mongo/backend/Utils.java

```diff
@@ -393,6 +402,10 @@ private static Object removeSubdocumentValue(Object document, String key, Atomic
             Object subObject = getFieldValueListSafe(document, mainKey);
             if (subObject instanceof Document || subObject instanceof List<?>) {
                 return removeSubdocumentValue(subObject, subKey, matchPos);
+            } else if (!isNullOrEmpty(subKey)) { // not missing, but not a Document or List, so no subdocuments
```

Comment:
I’m wondering if an empty subKey can even occur here…

Maybe we should simplify this if-else cascade to something like:
```java
            if (subObject instanceof Document || subObject instanceof List<?>) {
                return removeSubdocumentValue(subObject, subKey, matchPos);
            } else {
                return Missing.getInstance();
            }
```

Reply:
Yea, I thought about the same thing. However, that's drawing an implicit dependency that `getSubKey` will always do the right thing (return a subkey when a dot exists in the key). I figured it would be better to throw an exception if somehow `getSubKey` didn't do the right thing.

Thoughts?

(regardless, i'll change `subObject == Missing.getInstance()` to `subObject instanceof Missing`)

Reply:
Ok, good thoughts.

What do you think about adding an assertion (e.g. `Assert.isTrue(…)`) which makes sure that we would notice if `getSubKey` would change the behaviour in an unexpected/incompatible way?

Reply:
Ah sure, that's probably better.